### PR TITLE
if output is empty, auto return the error 404.

### PR DESCRIPTION
**Pull request by hyacinthus**
Hey,
    How about automatically return 404 when output data is empty?

### CODE REVIEW

**Comment by hfaran** [comment]
Hmm, I'm not quite certain if this is something we want to bake directly into `schema.validate` with no option to disable, since it's entirely possible someone may **want** an empty response from a request as defined behaviour (although this is unlikely).

That being said, I think it would be best to add a boolean keyword argument here like `on_empty_404=False` for example. Then if you wanted to use it by default, you could just create a partial in your application like so, and use it:

```python
import functools
validate = functools.partial(schema.validate, on_empty_404=True)
```

**Comment by hfaran** [comment]
Could you also add a test for this? Thanks!

**Comment by malexh** [comment]
There are definitely times when you may want an empty response.  For example, GET a list of user-uploaded images.  If you have a brand-new user and the request is GET images, the result should be empty to indicate they haven't uploaded any images yet.  Returning a 404 would be atypical behavior in this very standard use case.  Returning an empty set is expected.

**Comment by hyacinthus** [comment]
@hfaran @malexh  Agree with you.  Boolean keyword is better.  I'll test it.

**Comment by hyacinthus** [comment]
In fact I can't find a right place to add this feature. But It is stupid to check the output in every handler, or a decorator one more?  And now, I need import functools.partial, 2 lines per file, neither better nor worse...

**Comment by hfaran** [comment]
Well, you can just define a custom version of the decorator in your project like I do above, and then just import that.

For example, if you package is called `mywebapp`, just create a `schema` module and add this to it:
```python
import functools

from tornado_json import schema

validate = functools.partial(schema.validate, on_empty_404=True)
```

Then you can use `mywebapp.schema.validate` instead of `tornado_json.schema.validate`.

**Comment by hyacinthus** [comment]
Thanks , Good Idea .

File: tornado_json/schema.py [hunk 1]
```diff
@@ -75,6 +79,10 @@ def _wrapper(self, *args, **kwargs):
             #   we grab the output.
             if is_future(output):
                 output = yield output
+            
+            # if output is empty, auto return the error 404.
+            if not output and on_empty_404:
+                raise APIError(404,"Resource not found.")
```
Comment (hfaran) [inline tornado_json/schema.py]: should be a space after the comma

**Comment by hfaran** [comment]
@hyacinthus Are you planning on adding a test for this?

**Comment by hyacinthus** [comment]
Oh, sorry, I've misconstrued your words about test... I'll add it later.

File: tornado_json/schema.py [hunk 2]
```diff
@@ -15,7 +18,7 @@
 
 def validate(input_schema=None, output_schema=None,
              input_example=None, output_example=None,
-             format_checker=None):
+             format_checker=None, on_empty_404=False):
     """Parameterized decorator for schema validation
 
     :type format_checker: jsonschema.FormatChecker or None
```
Comment (hfaran) [inline tornado_json/schema.py]: Here can you also add:
```
:type on_empty_404: bool
:param on_empty_404: If this is set, and the result from the decorated method is a falsy value, a 404 will be raised.
```

File: tornado_json/schema.py [hunk 3]
```diff
@@ -38,6 +42,9 @@ def _validate(rh_method):
             or malformed
         :raises TypeError: If the output is invalid as per the schema
             or malformed
+        :raises APIError 404 not found: If the output is empty and 
+            on_empty_404 is True
+            warning: Number 0 is empty value in Python.
```
Comment (hfaran) [inline tornado_json/schema.py]: This should instead say:
```
:raises APIError: If the output is a falsy value and 
    on_empty_404 is True, an HTTP 404 error is returned
```

**Comment by hyacinthus** [comment]
OK

**Comment by hfaran** [comment]
Looks good! Merging.

**Comment by hyacinthus** [comment]
Thank you for your advice and patience.

**Pull request merged**